aten/src/ATen/core/Tensor.cpp

```diff
@@ -115,6 +303,19 @@ void Tensor::remove_hook(unsigned pos) const {
   impl::GetVariableHooks()->remove_hook(*this, pos);
 }
 
+bool Tensor::is_alias_of(const at::Tensor& other) const {
+  // If self and other are the same
+  if (unsafeGetTensorImpl() == other.unsafeGetTensorImpl()) return true;
+  // For tensors without storage, check alias_ information
+  if (has_view_meta() && other.has_view_meta()) {
+     return at::unsafeGetFunctionalImplBase(*this)->alias()->base().unsafeGetTensorImpl()
+         == at::unsafeGetFunctionalImplBase(other)->alias()->base().unsafeGetTensorImpl();
+  }
```

Comment:
So, this is an opportunity to fix the thing that XLA didn't do correctly.

XLA tensors don't correctly have a concept of storage or strides. This is a mistake. Although in functionalization tensors don't meaningfully have a concept of storage, we should still record accurate storage and strides in this case. In particular, you would allocate a storage whose data pointer is nullptr, but when a view operation occurs, you'd still ensure that the storages are shared. In other words, a FunctionalizeTensor should look "as if it were a normal tensor", including whether or not `is_alias_of` reports true (without this extra logic that was added here) and what `strides` reports. I did something similar when I added storages to meta tensors.

This is a performance pessimization: previously XLA/functionalize tensors didn't need to allocate storages, and now you're going to do that, and that's a dynamic allocation that's going to cost some refcount bumps / malloc hits, but IMO it's worth it. Furthermore, if we introduce a concept of StorageImpl subclass, you can actually pack Alias *directly into* the allocated storage (no need to have a separate field for it) and so you net out to zero allocations (well, technically the Alias is allocated lazily, but actually it is better not to do it lazily because of the multithreading problem identified earlier in the PR.)